packages/next/src/cli/next-test.ts

```diff
@@ -0,0 +1,205 @@
+  // Thus, `directory` could either be a valid Next.js project directory (that the user intends to run `next test` on), or it is the first argument for the test runner.
+  // Unfortunately, since many test runners support passing a path (to a test file or directory containing test files), we must check if `directory` is both a valid path and a valid Next.js project.
+
+  let baseDir, nextConfig
+
+  try {
+    // if directory is `undefined` or a valid path this will succeed.
+    baseDir = getProjectDir(directory, false)
+  } catch (err) {
+    // if that failed, then `directory` is not a valid path, so it must have meant to be the first item for `testRunnerArgs`
+    // @ts-expect-error directory is a string here since `getProjectDir` will succeed if its undefined
+    testRunnerArgs.unshift(directory)
+    // intentionally set baseDir to the resolved '.' path
+    baseDir = getProjectDir()
+  }
+
+  try {
+    // but, `baseDir` might not be a Next.js project directory, it could be a path-like argument for the test runner (i.e. `playwright test test/foo.spec.js`)
+    // if this succeeds, it means that `baseDir` is a Next.js project directory
+    nextConfig = await loadConfig(PHASE_PRODUCTION_BUILD, baseDir)
+  } catch (err) {
+    // if it doesn't, then most likely `baseDir` is not a Next.js project directory
+    // @ts-expect-error directory is a string here since `getProjectDir` will succeed if its undefined
+    testRunnerArgs.unshift(directory)
+    // intentionally set baseDir to the resolved '.' path
+    baseDir = getProjectDir()
+    nextConfig = await loadConfig(PHASE_PRODUCTION_BUILD, baseDir) // let this error bubble up if the `basePath` is still not a valid Next.js project
+  }
+
+  // set the test runner. priority is CLI option > next config > default 'playwright'
+  const configuredTestRunner =
+    options?.testRunner ?? // --test-runner='foo'
+    nextConfig.experimental.defaultTestRunner ?? // { experimental: { defaultTestRunner: 'foo' }}
+    'playwright'
+
+  if (!nextConfig.experimental.testProxy) {
+    return printAndExit(
+      `\`next experimental-test\` requires the \`experimental.testProxy: true\` configuration option.`
+    )
+  }
+
+  // execute test runner specific function
+  switch (configuredTestRunner) {
+    case 'playwright':
+      return runPlaywright(baseDir, nextConfig, testRunnerArgs)
+    default:
+      return printAndExit(
+        `Test runner ${configuredTestRunner} is not supported.`
+      )
+  }
+}
+
+async function checkRequiredDeps(
+  baseDir: string,
+  testRunner: SUPPORTED_TEST_RUNNERS
+) {
+  const deps = await hasNecessaryDependencies(
+    baseDir,
+    requiredPackagesByTestRunner[testRunner]
+  )
+  if (deps.missing.length > 0) {
+    await installDependencies(baseDir, deps.missing, true)
+
+    const playwright = spawn(
+      path.join(baseDir, 'node_modules', '@playwright', 'test', 'cli.js'),
```

Comment:
Is this the same pattern we use to spawn eslint from `next lint`? It should just be `pnpm playwright install`, `yarn playwright install` or `npx playwright install` depending on the package manager. Reaching into node_modules may not be supported in every package manager config and the path itself is internal and may change with any Playwright release.

Can be left as-is if we do the same for `next lint`

Reply:
next lint doesn't invoke `eslint` CLI, instead it uses its API. The implementation for `next test` is unique